### PR TITLE
upgrade Elasticsearch support to 8

### PR DESCRIPTION
**Pull request by tomkralidis**
# Overview
This PR upgrades pygeoapi Elasticsearch support to version 8, and drops support for ES 7 and earlier.
# Related Issue / Discussion
None
# Additional Information
None
# Contributions and Licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [x] I'd like to contribute [feature X|bugfix Y|docs|something else] to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution.
- [x] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

### CODE REVIEW

**Comment by tomkralidis** [comment]
@KoalaGeo any idea why the mainline CI did not run here?

**Comment by KoalaGeo** [comment]
> @KoalaGeo any idea why the mainline CI did not run here?

'lint workflow runs on all push and PR to `master`
`build` workflow only runs on the `master` branch
`containers` only runs when `build` completes

**Comment by webb-ben** [comment]
https://github.com/elastic/elasticsearch-dsl-py/issues/1569

File: pygeoapi/provider/elasticsearch_.py [hunk 1]
```diff
@@ -336,13 +307,15 @@ def query(self, offset=0, limit=10, resulttype='results',
...
+                matched = len(results['hits']['hits']) + offset
+                returned = len(results['hits']['hits'])
             else:
-                results = self.es.search(index=self.index_name,
-                                         from_=offset, size=limit,
-                                         body=query)
-                results['hits']['total'] = results['hits']['total']['value']
+                es_results = self.es.search(index=self.index_name,
+                                            from_=offset, size=limit, **query)
+                results = es_results.body
```
Comment (doublebyte1) [inline pygeoapi/provider/elasticsearch_.py:316]: The es_results structure no longer has a 'body' element:
`results = es_results.body` -> `results = es_results`

Comment (doublebyte1) [inline pygeoapi/provider/elasticsearch_.py:316]: The es_results structure no longer has a 'body' element:
`results = es_results.body` -> `results = es_results`

File: pygeoapi/provider/elasticsearch_.py [hunk 2]
```diff
@@ -336,13 +307,15 @@ def query(self, offset=0, limit=10, resulttype='results',
...
+                matched = es_results.body['hits']['total']['value']
```
Comment (doublebyte1) [inline pygeoapi/provider/elasticsearch_.py:317]: The same thing applies to this line:
`matched = es_results.body['hits']['total']['value']` -> `matched = es_results['hits']['total']['value']`

File: pygeoapi/provider/elasticsearch_.py [hunk 3]
```diff
@@ -399,12 +372,12 @@ def get(self, identifier, **kwargs):
             }
 
             try:
-                result = self.es.search(index=self.index_name, body=query)
-                if len(result['hits']['hits']) == 0:
+                result = self.es.search(index=self.index_name, **query)
+                if len(result.body['hits']['hits']) == 0:
```
Comment (doublebyte1) [inline pygeoapi/provider/elasticsearch_.py:376]: Remove body also here:
` if len(result.body['hits']['hits']) == 0:` -> ` if len(result['hits']['hits']) == 0:`

File: pygeoapi/provider/elasticsearch_.py [hunk 4]
```diff
@@ -399,12 +372,12 @@ def get(self, identifier, **kwargs):
...
                     LOGGER.error(err)
                     raise ProviderItemNotFoundError(err)
                 LOGGER.debug('Serializing feature')
-                feature_ = self.esdoc2geojson(result['hits']['hits'][0])
+                feature_ = self.esdoc2geojson(result.body['hits']['hits'][0])
```
Comment (doublebyte1) [inline pygeoapi/provider/elasticsearch_.py:380]: ` feature_ = self.esdoc2geojson(result.body['hits']['hits'][0])` ->  `feature_ = self.esdoc2geojson(result['hits']['hits'][0])`

**Review by doublebyte1 — COMMENTED** [review]
1. Removing the 'body" element, mentioned in the comments should allow these PR to pass the elasticsearch provider tests.
2. If we use the latest elastic client, we no longer need to run the ES container in compatibility mode. This [block](https://github.com/geopython/pygeoapi/blob/master/docker/examples/elastic/docker-compose.yml#L65) can be removed from the docker-compose.

File: pygeoapi/provider/elasticsearch_.py (hunk 1 above)
Comment (tomkralidis) [reply pygeoapi/provider/elasticsearch_.py]: Fixed.

File: pygeoapi/provider/elasticsearch_.py (hunk 2 above)
Comment (tomkralidis) [reply pygeoapi/provider/elasticsearch_.py]: Fixed.

File: pygeoapi/provider/elasticsearch_.py (hunk 3 above)
Comment (tomkralidis) [reply pygeoapi/provider/elasticsearch_.py]: Fixed.

File: pygeoapi/provider/elasticsearch_.py (hunk 4 above)
Comment (tomkralidis) [reply pygeoapi/provider/elasticsearch_.py]: Fixed.

**Comment by tomkralidis** [comment]
FYI this PR is now ready.  I am keeping as WIP to sort out the following issues with Python package dependencies:

- elasticsearch-dsl is still bound to `<8` while elasticsearch supports 8+.  Installing elasticsearch-dsl after elasticsearch will downgrade the latter to `<8`
- PostgreSQL tests fail given SQLAlchemy 2 which is installed via pygeofilter

As a result, in CI, I have done the following as workarounds AFTER the main `requirements-*txt` files are installed in `.github/workflows/main.yml`:
- upgrade `elasticsearch`
- downgrade `SQLAlchemy`

Questions:
- @francbartoli is elasticsearch-dsl something we will support long term?  See https://github.com/elastic/elasticsearch-dsl-py/issues/1569
- @KoalaGeo should we update the PostgreSQL plugin (in another PR)?

cc @kalxas

**Comment by tomkralidis** [comment]
Proposing the following:
- merge PR on approval
- open an issue on elasticsearch-dsl support
- open an issue on SQLAlchemy 2 support

Thoughts @francbartoli @kalxas @doublebyte1 @webb-ben ?

**Comment by tomkralidis** [comment]
Issues filed:
- https://github.com/geopython/pygeoapi/issues/1123
- https://github.com/geopython/pygeoapi/issues/1124

**Review by francbartoli — APPROVED** [review]
I would manage the issues separately in the tickets which have been opened

**Comment by kalxas** [comment]
I agree with @francbartoli

**Review by kalxas — APPROVED** [review]

**Comment by francbartoli** [comment]
Maybe we could cut a release before merging this, what do you think @kalxas?

**Review by webb-ben — APPROVED** [review]
+1 from me. 

Is it worth adding documentation to clarify the sequence to install correct package versions? or is that covered by the pending issues?

File: docs/source/data-publishing/ogcapi-features.rst [hunk 5]
```diff
@@ -73,11 +73,11 @@ Elasticsearch
 ^^^^^^^^^^^^^
 
 .. note::
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (tomkralidis) [reply docs/source/data-publishing/ogcapi-features.rst]: Good catch. Fixed.

**Comment by tomkralidis** [comment]
Thinking more, the safest bet is cutting a release prior to merging this PR.  We can determine what kind of documentation is needed at that point depending on `elasticsearch-dsl` status and our use of same.

**Comment by kalxas** [comment]
ok, a new release is now done.
Also the SQLAlchemy issue is fixed on master

**Pull request merged**